internal/lint/lint.go

```diff
@@ -0,0 +1,335 @@
+	GroupToCheckers = map[string][]Checker{
+		DefaultGroup: DefaultCheckers,
+		AllGroup:     AllCheckers,
+	}
+)
+
+func init() {
+	ids := make(map[string]struct{})
+	for _, checker := range AllCheckers {
+		if _, ok := ids[checker.ID()]; ok {
+			panic(fmt.Sprintf("duplicate checker id %s", checker.ID()))
+		}
+		ids[checker.ID()] = struct{}{}
+	}
+}
+
+// Runner runs a lint job.
+type Runner interface {
+	Run(...*file.ProtoSet) ([]*text.Failure, error)
+}
+
+// RunnerOption is an option for a new Runner.
+type RunnerOption func(*runner)
+
+// RunnerWithLogger returns a RunnerOption that uses the given logger.
+//
+// The default is to use zap.NewNop().
+func RunnerWithLogger(logger *zap.Logger) RunnerOption {
+	return func(runner *runner) {
+		runner.logger = logger
+	}
+}
+
+// NewRunner returns a new Runner.
+func NewRunner(options ...RunnerOption) Runner {
+	return newRunner(options...)
+}
+
+// The below should not be needed in the CLI
+// TODO make private
+
+// Checker is a linter for Protobuf files.
+type Checker interface {
+	// Return the ID of this Checker. This should be all UPPER_SNAKE_CASE.
+	ID() string
+	// Return the purpose of this Checker. This should be a human-readable string.
+	Purpose() string
+	// Check the file data for the descriptors in a common directgory.
+	// If there is a lint failure, this returns it in the
+	// slice and does not return an error. An error is returned if something
+	// unexpected happens. Callers should verify the files are compilable
+	// before running this.
+	Check(dirPath string, descriptors []*proto.Proto) ([]*text.Failure, error)
+}
+
+// NewChecker is a convenience function that returns a new Checker for the
+// given parameters.
+//
+// The ID will be upper-cased.
+//
+// Failures returned from check do not need to set the ID, this will be overwritten.
+func NewChecker(id string, purpose string, check func(string, []*proto.Proto) ([]*text.Failure, error)) Checker {
+	return newBaseChecker(id, purpose, check)
+}
+
+// NewAddChecker is a convenience function that returns a new Checker for the
+// given parameters, using a function to record failures.
+//
+// The ID will be upper-cased.
+//
+// Failures returned from check do not need to set the ID, this will be overwritten.
+func NewAddChecker(id string, purpose string, addCheck func(func(*text.Failure), string, []*proto.Proto) error) Checker {
+	return newBaseAddChecker(id, purpose, addCheck)
+}
+
+// GetCheckers returns the Checkers for the LintConfig.
+//
+// The config is expected to be valid, ie slices deduped, all upper-case,
+// and only either IDs or Group/IncludeIDs/ExcludeIDs, with no overlap between
+// IncludeIDs and ExcludeIDs.
+//
+// If the config came from the settings package, this is already validated.
+func GetCheckers(config settings.LintConfig) ([]Checker, error) {
+	if len(config.IDs) == 0 && (len(config.Group) == 0 || config.Group == DefaultGroup) && len(config.IncludeIDs) == 0 && len(config.ExcludeIDs) == 0 {
```

Comment:
We might want to make a simple helper function for these types of conditionals:
```
if config.IsEmpty() {
  return DefaultCheckers, nil
}
```

Reply:
`IsEmpty` might not be the best name here.

Reply:
I think I did that originally, but it ended up being weirder to read than this.